Modification related to EBS data volume replacement (#195)
scalar-labs/scalar-terraform: docs/CassandraOperation.md

JVM_OPTS="$JVM_OPTS -Dcassandra.replace_address_first_boot=<dead_node_ip>"
```

> You also need to modify the file `/etc/cassandra/conf/cassandra.yaml` If you replaced a seed node and a volume.

Reviewer: ```suggestion
You also need to modify the file `/etc/cassandra/conf/cassandra.yaml` if you replaced a seed node and a volume.
```

* Remove the newly created Cassandra node IP from the seeds of the node.

Finally you can start the Cassandra service.

```console
sudo systemctl start cassandra
```

> If you replaced a seed node, you should replace the seeds of all Cassandra nodes with newly created Cassandra node IP.

Reviewer: ```suggestion
If you replaced a seed node (old seed node), you should replace the old seed node's IP of all Cassandra nodes with newly created Cassandra node IP.
```

JVM_OPTS="$JVM_OPTS -Dcassandra.replace_address_first_boot=<dead_node_ip>"
```

* You need to modify the file `/etc/cassandra/conf/cassandra.yaml` If you choose *Taint Volume*

>   * Remove the newly created cassandra node IP from the newly created cassandra node seeds.

Reviewer: ```suggestion
* Remove the newly created Cassandra node IP from the seeds of the node.
```

>   * Remove the newly created cassandra node IP from the newly created cassandra node seeds.

Reviewer 2: [Q] just remove ip from the newly created node? or all each nodes if replace node is seed node?

https://docs.datastax.com/en/ddac/doc/datastax_enterprise/operations/opsReplaceNode.html

Author: @tei-k san,
> just remove ip from the newly created node? or all each nodes if replaced node was seed node?

Before restarting the newly created node, We need to remove the newly created node IP from the `seeds` of newly created node.
Note:- It is applicable only for seed nodes. We can skip this step when the newly created node is a `non-seed` node.

Reviewer 3: Just curious. Is there another seed in this case? Can a new node join the cluster without removing the IP from its `seeds` parameter if other seeds are alive?

Author: > Is there another seed in this case? 

Yes, In this case, other two `seed` nodes are available.

> Can a new node join the cluster without removing the IP from its seeds parameter if other seeds are alive?

Do you mean, 
Is it possible to add a new node in the cluster while one seed node crashed in the cluster?

Reviewer 3: > Is it possible to add a new node in the cluster while one seed node crashed in the cluster?
No. Can a new node with which the seed node was replaced join to the cluster without removing the seed's IP?
You removed the seed's IP from the config file of the new node in your test. I'd like to know what happens in the bootstrapping when it is NOT removed.

Author: > Can a new node with which the seed node was replaced join to the cluster without removing the seed's IP? 

yes, But both nodes will show in the cluster. (Old node status is `DN`)

> I'm wondering whether the new node checks another seed or not.
You removed the seed's IP from the config file of the new node in your test. I'd like to know what happens in the bootstrapping when it is NOT removed.

Bootstrapping will happens while the seeds are not replaced.

Reviewer 3: The terraform module automatically adds the new node's IP to `seeds` in the config file. So, we need to remove the IP from the `seeds` explicitly and restart the new node with `replace_address_first_boot` option. I got it.

> * Finally you can start the Cassandra service.

Reviewer: ```suggestion
Finally you can start the Cassandra service.
```

```console
sudo systemctl start cassandra
```

> * After a successful restart of the newly created Cassandra node, you should replace the seeds of all Cassandra nodes with newly created Cassandra node IP and restart the Cassandra nodes.

Author: > What do you mean by this?

All other Cassandra node seeds contain IP address of the old node(crashed node) When we replace a seed node. So we need to replace the IP address in all other Cassandra node seeds.

> I think non replaced nodes don't have the new node IP in the seeds list, doesn't it?

yes, We do not need to update the seeds for non replaced nodes

> * After a successful restart of the newly created Cassandra node, you should replace the seeds of all Cassandra nodes with newly created Cassandra node IP and restart the Cassandra nodes.

Author: Thank you for the review comment.
Document updated with the above change.

> * After a successful restart of the newly created Cassandra node, you should replace the seeds of all Cassandra nodes with newly created Cassandra node IP and restart the Cassandra nodes.

Reviewer 3: I think that we don't need to make the new node a seed node. The new node might not be able to be accessed because the new node will have a heavy load to sync the data of the cluster.

I have two scenarios to replace a seed node. I'm not sure which case you tested.
### A) A cluster has only one seed node. The seed node is dead and we need to replace the node.
- The cluster has no living seed node, so, a new node can't join the cluster.
- Before replacing, we need to promote an existing node to a seed node. (We can remove the dead seed node's IP from other nodes in this promotion). After this promotion, this case will become a case where replacing a non-seed node.

### B) A cluster has multiple seed nodes. One of the seed nodes is dead and we replace it.
- A new node can join the cluster as a non-seed node.

According to the [document](https://docs.datastax.com/en/cassandra-oss/2.1/cassandra/operations/opsReplaceNode.html#Replacingadeadseednode), DataStax recommends promoting a non-seed node before replacing the dead seed node.

> * After a successful restart of the newly created Cassandra node, you should replace the seeds of all Cassandra nodes with newly created Cassandra node IP and restart the Cassandra nodes.

Reviewer 3: Sorry, the document is too old.
The current [document](https://docs.datastax.com/en/cassandra-oss/3.x/cassandra/operations/opsReplaceNode.html) says that a new node joins the cluster as a new seed node.

```console

> sudo systemctl restart cassandra

Reviewer: ```suggestion
```
